### ISSUE TITLE
Server_relay disabled, prevents mastersync from puppet function to call master

### ISSUE THREAD

**Issue by ThaBalla1148** [godotengine/godot#37996]
**Godot version:**
3.2.1.stable.official

**OS/device including version:**
Windows 10 Pro 18362.720

**Issue description:**
With the server's `server_relay` flag set to false, a `puppet` function does not properly invoke a `mastersync` function on the server, only locally. The client is using `rpc()`. From my understanding a `mastersync` function should always be called on the `master` and run locally on the caller. Why are these calls not making it to the server when being invoked within a `puppet` function?

Also, with the server's `server_relay` flag set to true, the `puppet` functions are properly invoked on the server, but it is also relayed to the other clients, which is what I want to avoid due to loads of errors, design, and security concerns. Why would a `mastersync` function call get relayed to other puppets in this case?

These behaviors do not seem to be consistent under these circumstances.

**Steps to reproduce:**
- Create a Server with `server_relay` set to false
```
func create_server():
	var host = NetworkedMultiplayerENet.new()
	host.server_relay = false
	host.create_server(DEFAULT_PORT, MAX_PLAYERS)
	get_tree().set_network_peer(host)
```
- Create a Client
```
func connect_to_server():
	var host = NetworkedMultiplayerENet.new()
	host.create_client(ip, DEFAULT_PORT)
	get_tree().set_network_peer(host)
```
- Create a `mastersync` function on the Server and Client
```
mastersync func test(var network_id, var val: int):
	print_debug("Test called from: ", network_id, " ", val)
	print_debug("Test running on: ", get_tree().get_network_unique_id(), " ", val)
```
- Create Client signal function and connect to `connected_to_server`
```
func _ready():
	get_tree().connect("connected_to_server", self, "_connected_ok")

func _connected_ok():
	rpc("test", get_tree().get_network_unique_id(), 1) # This works properly
```
- Create a `puppet` function on the Client that invokes the `mastersync` function
```
puppet func pre_start_game():
	rpc("test", get_tree().get_network_unique_id(), 2) # This does not work properly
```
- Run a Server and 2 Clients
- On the server, call `rpc()` to invoke the puppet function for all clients/puppets
```
rpc("pre_start_game")
```

**Results**
Server prints:
```
Test called from: 1149837705 1
Test running on: 1 1
Test called from: 111413756 1
Test running on: 1 1
```

Client A prints:
```
Test called from: 1149837705 1
Test running on: 1149837705 1
Test called from: 1149837705 2
Test running on: 1149837705 2
```

Client B prints:
```
Test called from: 111413756 1
Test running on: 111413756 1
Test called from: 111413756 2
Test running on: 111413756 2
```
**Minimal reproduction project:**
The provided projects were repurposed to only test the functionality of making an `rpc()` call from a `puppet` function to a `mastersync` function with `server_relay` disabled.

[Godot.3.2.1-puppet-mastersync-test.zip](https://github.com/godotengine/godot/files/4497612/Godot.3.2.1-puppet-mastersync-test.zip)

**Comment by jonbonazza** [comment]
I don't understand. It looks like it's working correctly to me. What do you expect to see?

**Comment by briansemrau** [comment]
I've reproduced this with a simpler reproduction (no need for puppet function). The RPC call from the client is received by the server exactly once, and no further calls are received.

The following code produces the issue:
```gdscript
func create_server():
	var host = NetworkedMultiplayerENet.new()
	host.server_relay = false
	host.create_server(29264)
	get_tree().set_network_peer(host)
	print("Started server.")

func connect_to_server():
	var host = NetworkedMultiplayerENet.new()
	host.create_client("127.0.0.1", 29264)
	get_tree().set_network_peer(host)
	print("Joined server.")

func call_knock():
	rpc("_knock", "from {0}".format([get_tree().get_network_unique_id()]))

mastersync func _knock(message):
	print("Knocking! {0}".format([message]))
```
Outputs showcasing incorrect behavior:
```
Started server.
Knocking! from 1
Knocking! from 1
Knocking! from 467961845
```
```
Joined server.
Knocking! from 467961845
Knocking! from 467961845
Knocking! from 467961845
Knocking! from 467961845
Knocking! from 467961845
```

This issue no longer occurs when using either
- `host.server_relay = true` OR
- rpc_id instead of rpc: `rpc_id(1, "_knock", "from {0}".format([get_tree().get_network_unique_id()]))`

Outputs behaving as expected:
```
Started server.
Knocking! from 1
Knocking! from 1
Knocking! from 2118142965
Knocking! from 2118142965
```
```
Joined server.
Knocking! from 2118142965
Knocking! from 2118142965
```

Hopefully this makes the issue clear.

Min rep:
[TestRelayIssue3x.zip](https://github.com/godotengine/godot/files/6253189/TestRelayIssue3x.zip)

**Comment by briansemrau** [comment]
> 
> 
> I don't understand. It looks like it's working correctly to me. What do you expect to see?

From the authors example, the server prints:
```
Test called from: 1149837705 1
Test running on: 1 1
Test called from: 111413756 1
Test running on: 1 1
```
when I expect it should be printing:
```
Test called from: 1149837705 1
Test running on: 1 1
Test called from: 111413756 1
Test running on: 1 1
Test called from: 1149837705 2
Test running on: 1 2
Test called from: 111413756 2
Test running on: 1 2
```

**Comment by briansemrau** [comment]
I just tested and confirmed this on latest 4.0 ed2f51b15f as well.

**Comment by jordo** [comment]
@briansemrau  can you try rpc_id with target peer server?

**Comment by jordo** [comment]
`rpc_id` instead of `rpc` and use the target peer TARGET_PEER_SERVER ?

**Comment by briansemrau** [comment]
> 
> 
> can you try rpc_id with target peer server?
> `rpc_id` instead of `rpc` and use the target peer TARGET_PEER_SERVER ?

@jordo
Using `rpc_id(TARGET_PEER_SERVER, ...)` with a `mastersync` function behaves the same as if it were just a `master` function. It only executes on the server when I want it to execute on both client and server.

The issue is with TARGET_PEER_BROADCAST. With `server_relay=false`, I expect the `rpc` call to execute the function on both the client and the server every time, but in actuality it executes on the client every time but only once on the server.

**Comment by jordo** [comment]
Hmm, the once thing is odd...  Perhaps it's a `mastersync` issue.  But i think @Faless is in support of https://github.com/godotengine/godot/pull/38030  .  From an api standpoint i believe it makes sense to have server included in broadcast peer regardless of the relay mode for the server.

**Comment by briansemrau** [comment]
> 
> 
> @briansemrau Can you try with this patch? #38030

I merged it locally against 4.0 ed2f51b15f and confirmed that it does resolve the issue. 👍
